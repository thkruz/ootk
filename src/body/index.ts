export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
