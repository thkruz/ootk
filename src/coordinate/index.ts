export * from './Hill.js';
export * from './ClassicalElements.js';
export * from './EquinoctialElements.js';
export * from './FormatTle.js';
export * from './Geodetic.js';
export * from './ITRF.js';
export * from './J2000.js';
export * from './RelativeState.js';
export * from './RIC.js';
export * from './StateVector.js';
export * from './TEME.js';
export * from './Tle.js';
