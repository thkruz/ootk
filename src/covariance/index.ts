export * from './CovarianceSample.js';
export * from './StateCovariance.js';
