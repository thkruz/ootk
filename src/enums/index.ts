export { CatalogSource } from './CatalogSource.js';
export { CommLink } from './CommLink.js';
export { OrbitRegime } from './OrbitRegime.js';
export { Sgp4OpsMode } from './Sgp4OpsMode.js';
export { AngularDiameterMethod } from './AngularDiameterMethod.js';
export { AngularDistanceMethod } from './AngularDistanceMethod.js';
export { PassType } from './PassType.js';
