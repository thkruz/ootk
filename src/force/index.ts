export { EarthGravity } from './EarthGravity.js';
export { Force } from './Force.js';
export { ForceModel } from './ForceModel.js';
export { Gravity } from './Gravity.js';
export { SolarRadiationPressure } from './SolarRadiationPressure.js';
export { ThirdBodyGravity } from './ThirdBodyGravity.js';
export { Thrust } from './Thrust.js';

