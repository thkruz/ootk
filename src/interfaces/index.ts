export type { BaseObjectParams } from './BaseObjectParams.js';
export type { GroundPositionParams } from './GroundPositionParams.js';
export type { OptionsParams } from './OptionsParams.js';
export type { SatelliteParams } from './SatelliteParams.js';
export type { SensorParams } from './SensorParams.js';
export type { StarObjectParams } from './StarObjectParams.js';
export type { ClassicalElementsParams } from './ClassicalElementsParams.js';
export type { EquinoctialElementsParams } from './EquinoctialElementsParams.js';
export type { RfSensorParams } from './RfSensorParams.js';
