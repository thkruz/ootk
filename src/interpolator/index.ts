export { ChebyshevCoefficients } from './ChebyshevCoefficients.js';
export { ChebyshevInterpolator } from './ChebyshevInterpolator.js';
export { CubicSpline } from './CubicSpline.js';
export { CubicSplineInterpolator } from './CubicSplineInterpolator.js';
export { Interpolator } from './Interpolator.js';
export { LagrangeInterpolator } from './LagrangeInterpolator.js';
export { StateInterpolator } from './StateInterpolator.js';
export { VerletBlendInterpolator } from './VerletBlendInterpolator.js';

