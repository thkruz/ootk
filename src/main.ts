/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
