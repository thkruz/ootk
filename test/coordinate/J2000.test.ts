// Generated by CodiumAI

import { EpochUTC, ClassicalElements, Kilometers, Radians, J2000, Vector3D, KilometersPerSecond } from '../../src/main';

describe('J2000', () => {
  const exampleDate = new Date(1705109326817);
  const epoch = EpochUTC.fromDateTime(exampleDate);

  // can be created from classical elements
  it('should create a J2000 coordinate from classical elements', () => {
    const elements = new ClassicalElements({
      epoch,
      semimajorAxis: 6943.547853722985 as Kilometers,
      eccentricity: 0.0011235968124658146,
      inclination: 0.7509087232045765 as Radians,
      rightAscension: 0.028239555738616327 as Radians,
      argPerigee: 2.5386411901807353 as Radians,
      trueAnomaly: 0.5931399364974058 as Radians,
    });
    const j2000 = J2000.fromClassicalElements(elements);

    expect(j2000.epoch).toEqual(elements.epoch);
    expect(j2000.position).toEqual(elements.toPositionVelocity().position);
    expect(j2000.velocity).toEqual(elements.toPositionVelocity().velocity);
  });

  // can get the name of the coordinate system
  it('should return the name of the coordinate system as "J2000"', () => {
    const j2000 = new J2000(
      epoch,
      new Vector3D<Kilometers>(5000 as Kilometers, 10000 as Kilometers, 2100 as Kilometers),
      new Vector3D<KilometersPerSecond>(7 as KilometersPerSecond, 4 as KilometersPerSecond, 2 as KilometersPerSecond),
    );

    expect(j2000.name).toBe('J2000');
  });

  // can get a value indicating whether the coordinate system is inertial
  it('should return true for the inertial property', () => {
    const j2000 = new J2000(
      epoch,
      new Vector3D<Kilometers>(5000 as Kilometers, 10000 as Kilometers, 2100 as Kilometers),
      new Vector3D<KilometersPerSecond>(7 as KilometersPerSecond, 4 as KilometersPerSecond, 2 as KilometersPerSecond),
    );

    expect(j2000.inertial).toBe(true);
  });

  // Check conversion to and from
  it('should convert to and from other coordinate systems', () => {
    const j2000 = new J2000(
      epoch,
      new Vector3D<Kilometers>(5000 as Kilometers, 10000 as Kilometers, 2100 as Kilometers),
      new Vector3D<KilometersPerSecond>(7 as KilometersPerSecond, 4 as KilometersPerSecond, 2 as KilometersPerSecond),
    );
    const itrf = j2000.toITRF();
    const j200FromItf = itrf.toJ2000();

    expect(j200FromItf.position.x).toBeCloseTo(j2000.position.x, 8);
    expect(j200FromItf.position.y).toBeCloseTo(j2000.position.y, 8);
    expect(j200FromItf.position.z).toBeCloseTo(j2000.position.z, 8);

    const teme = j2000.toTEME();
    const j200FromTeme = teme.toJ2000();

    expect(j200FromTeme.position.x).toBeCloseTo(j2000.position.x, 8);
    expect(j200FromTeme.position.y).toBeCloseTo(j2000.position.y, 8);
    expect(j200FromTeme.position.z).toBeCloseTo(j2000.position.z, 8);
  });

  // toString
  it('should return a string', () => {
    const j2000 = new J2000(
      epoch,
      new Vector3D<Kilometers>(5000 as Kilometers, 10000 as Kilometers, 2100 as Kilometers),
      new Vector3D<KilometersPerSecond>(7 as KilometersPerSecond, 4 as KilometersPerSecond, 2 as KilometersPerSecond),
    );

    expect(j2000.toString()).toMatchSnapshot();
  });

  // mechanical energy
  it('should calculate the mechanical energy', () => {
    const j2000 = new J2000(
      epoch,
      new Vector3D<Kilometers>(5000 as Kilometers, 10000 as Kilometers, 2100 as Kilometers),
      new Vector3D<KilometersPerSecond>(7 as KilometersPerSecond, 4 as KilometersPerSecond, 2 as KilometersPerSecond),
    );

    expect(j2000.mechanicalEnergy).toMatchSnapshot();
  });

  // period
  it('should calculate the period', () => {
    const j2000 = new J2000(
      epoch,
      new Vector3D<Kilometers>(5000 as Kilometers, 10000 as Kilometers, 2100 as Kilometers),
      new Vector3D<KilometersPerSecond>(7 as KilometersPerSecond, 4 as KilometersPerSecond, 2 as KilometersPerSecond),
    );

    expect(j2000.period).toMatchSnapshot();
  });

  // angular rate
  it('should calculate the angular rate', () => {
    const j2000 = new J2000(
      epoch,
      new Vector3D<Kilometers>(5000 as Kilometers, 10000 as Kilometers, 2100 as Kilometers),
      new Vector3D<KilometersPerSecond>(7 as KilometersPerSecond, 4 as KilometersPerSecond, 2 as KilometersPerSecond),
    );

    expect(j2000.angularRate).toMatchSnapshot();
  });

  // toClassicalElements
  it('should convert to classical elements', () => {
    const j2000 = new J2000(
      epoch,
      new Vector3D<Kilometers>(5000 as Kilometers, 10000 as Kilometers, 2100 as Kilometers),
      new Vector3D<KilometersPerSecond>(7 as KilometersPerSecond, 4 as KilometersPerSecond, 2 as KilometersPerSecond),
    );
    const elements = j2000.toClassicalElements();

    expect(elements.epoch).toEqual(epoch);
    expect(elements.semimajorAxis).toMatchSnapshot();
    expect(elements.eccentricity).toMatchSnapshot();
    expect(elements.inclination).toMatchSnapshot();
    expect(elements.rightAscension).toMatchSnapshot();
    expect(elements.argPerigee).toMatchSnapshot();
    expect(elements.trueAnomaly).toMatchSnapshot();
  });
});
