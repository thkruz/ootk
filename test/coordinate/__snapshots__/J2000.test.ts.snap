// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`J2000 should calculate the angular rate 1`] = `0.000002809314397474213`;

exports[`J2000 should calculate the mechanical energy 1`] = `-0.539173763730993`;

exports[`J2000 should calculate the period 1`] = `37275.911593878845`;

exports[`J2000 should convert to classical elements 1`] = `369640.0569843672`;

exports[`J2000 should convert to classical elements 2`] = `0.9909435374737835`;

exports[`J2000 should convert to classical elements 3`] = `2.896313029154172`;

exports[`J2000 should convert to classical elements 4`] = `1.955753713088999`;

exports[`J2000 should convert to classical elements 5`] = `5.144903601262446`;

exports[`J2000 should convert to classical elements 6`] = `2.0019281534313316`;

exports[`J2000 should return a string 1`] = `
"[J2000]
  Epoch: 2024-01-13T01:28:46.817Z
  Position: [5000.000000, 10000.000000, 2100.000000] km
  Velocity: [7.000000000, 4.000000000, 2.000000000] km/s"
`;
